Sticky target thread_limit implementation

sandeepkosuri opened the pull request:
- Initial implementation for `target` directive's `thread_limit` clause
- In this implementation, All the `parallel` based directives gets a `num_threads` clause if it's in a `target` directive scope which has `thread_limit` defined.

Commit:
Sticky target thread_limit implementation

sandeepkosuri commented:
- I will try to eliminate the redundant creation of `num_threads_clause` in this implementation.
- I noticed a flaw, with this implementation, the `thread_limit` set by parent `target` overrides the `num_threads` clauses set by child `parallel` directives. We don't want that to happen if the `num_threads` of `parallel` is lower than the `thread_limit` of `target`. I will fix it soon.

soumitrachatterjee commented on clang/include/clang/Sema/Scope.h:
```diff
@@ -236,12 +236,25 @@ class Scope {
 
   void setFlags(Scope *Parent, unsigned F);
 
+  unsigned TargetThreadLimit = 0;
```
This will probably need to change (along with all subsequent uses) to support generic expressions and not just constants.

```suggestion
  clang::Expr TargetThreadLimit = nullptr;
```

soumitrachatterjee commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -2909,6 +2909,38 @@ StmtResult Parser::ParseOpenMPDeclarativeOrExecutableDirective(
       if (Clause) {
+        // in the case of target directive, set the thread_limit for this scope
+        if (DKind == OMPD_target && CKind == OMPC_thread_limit) {
+          IntegerLiteral *TL = nullptr;
```
Note that `IntegerLiteral` is already a derivative of `clang::Expr` and hence support for generic expressions should not require substantial changes.

soumitrachatterjee commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -2909,6 +2909,38 @@ StmtResult Parser::ParseOpenMPDeclarativeOrExecutableDirective(
+            TL = dyn_cast<IntegerLiteral>(i);
+            break;
+          }
+          unsigned TL_val = TL->getValue().getSExtValue();
+          if (TL_val) {
+            Actions.getCurScope()->setTargetThreadLimit(TL_val);
+          }
```
In order to support generic expressions, you would not extract the value of the thread limit, but instead use it as is in the form of `clang::Expr`.

soumitrachatterjee commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -2909,6 +2909,38 @@ StmtResult Parser::ParseOpenMPDeclarativeOrExecutableDirective(
+            clang::isOpenMPParallelDirective(DKind)) {
+          if (unsigned target_thread_limit =
+                  Actions.getCurScope()->getParent()->getParentThreadLimit()) {
+            llvm::APInt TL_val =
+                llvm::APInt(/*numBits*/ 32, target_thread_limit);
```
This again simplifies in case of generic expressions and you can use the expression directly instead of extracting the value and then creating an expression explicitly.

soumitrachatterjee commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -2909,6 +2909,38 @@ StmtResult Parser::ParseOpenMPDeclarativeOrExecutableDirective(
+                  Actions.getCurScope()->getParent()->getParentThreadLimit()) {
+            llvm::APInt TL_val =
+                llvm::APInt(/*numBits*/ 32, target_thread_limit);
+            IntegerLiteral *TL = IntegerLiteral::Create(
+                Actions.getASTContext(), TL_val,
+                Actions.getASTContext().getIntTypeForBitwidth(32, /*Signed=*/0),
+                Loc);
```
Instead of creating this, you would simply reuse the saved expression.

soumitrachatterjee commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -2918,6 +2950,29 @@ StmtResult Parser::ParseOpenMPDeclarativeOrExecutableDirective(
+      OMPClause *num_threads_clause;
+      if (unsigned target_thread_limit =
+              Actions.getCurScope()->getParent()->getParentThreadLimit()) {
+        llvm::APInt TL_val = llvm::APInt(/*numBits*/ 32, target_thread_limit);
+        IntegerLiteral *TL = IntegerLiteral::Create(
+            Actions.getASTContext(), TL_val,
+            Actions.getASTContext().getIntTypeForBitwidth(32, /*Signed=*/0),
+            Loc);
```
Same here as well.

soumitrachatterjee commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -2918,6 +2950,29 @@ StmtResult Parser::ParseOpenMPDeclarativeOrExecutableDirective(
+    // 'target thread_limit()' scope
+    if (clang::isOpenMPParallelDirective(DKind) &&
+        !FirstClauses[unsigned(OMPC_num_threads)].getInt()) {
+      OMPClause *num_threads_clause;
```
Must initialize to NULL!
```suggestion
      OMPClause *num_threads_clause = nullptr;
```

soumitrachatterjee commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -2918,6 +2950,29 @@ StmtResult Parser::ParseOpenMPDeclarativeOrExecutableDirective(
+            Actions.ActOnOpenMPNumThreadsClause(TL, Loc, Loc, Loc);
+      }
+      FirstClauses[unsigned(OMPC_num_threads)].setInt(true);
+      if (num_threads_clause) {
```
This check is bad if `num_threads_clause` has not been initialized as suggested above.

Commit:
Code refactoring

This commit,
- Adds checks before casts
- Compares num_threads and thread_limit and chooses the loweset one
- Moves some code to a different function for readability

With the current implementation, target's thread_limit is
honored iff argument to thread_limit is a constant expr.
It also fails to affect an enclosed directive's num_threads clause
if the clause's argument is anything other than a constant expr.

TODO: support general expression arguments to thread_limit and
      num_threads

Commit:
Initial support for general expressions

This commit adds support for general expressions as argument for
target thread_limit(). But the functionality of choosing the lower
limit when there is a conflicting num_threads() clause inside the
target thread_limit region is removed as it can't be done at parse
phase anymore.

TODO: support the cases where there is a conflicting num_threads
      clause in enclosed parallel directives.

Commit:
uses fork_teams() to call offloading func

Commit:
shifting changes to CGOpenMPRuntime.cpp

Commit:
code refactoring and emit num_teams

alexey-bataev commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -9859,7 +9859,18 @@ void CGOpenMPRuntime::emitTargetCall(
+
+          emitNumTeamsClause(CGF, NumTeams, ThreadLimit, D.getBeginLoc());
+        }
+        emitTeamsCall(CGF, D, D.getBeginLoc(), OutlinedFn, CapturedVars);
```
Why emitTeamsCall here if it is only target region?

sandeepkosuri replied:
I was using `emitNumTeamsClause()` to set the `thread_limit` of the target region. As this call would effect only one consequent `fork` call, I used `emitTeamsCall()` to call the offloading function which holds the target region instructions. With this method my IR looks something like this:
```
define dso_local i32 @main() #0 {
…
call void @__kmpc_push_num_teams(ptr @1, i32 %0, i32 0, i32 %3)
…
call void (ptr, i32, ptr, ...) @__kmpc_fork_teams(ptr @1, i32 2, ptr @__omp_offloading_fd06_8039e46d_main_l15, i64 %5, i64 %7)
…
}
```
This sets the thread_limit for target region correctly. But this approach is making the variables defined outside the target region to not be firstprivate in target region.

alexey-bataev replied:
And it means it is not correct, I assume. Need to find a better approach.

sandeepkosuri replied:
Yeah. I am working on a solution which tries to set thread_limit icv at runtime, I will upload that soon, in this same PR.

alexey-bataev commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -9859,7 +9859,18 @@ void CGOpenMPRuntime::emitTargetCall(
       }
-      emitOutlinedFunctionCall(CGF, D.getBeginLoc(), OutlinedFn, CapturedVars);
+      if (D.getDirectiveKind() == OMPD_target && D.getSingleClause<OMPThreadLimitClause>()) {
+        const auto *TL = D.getSingleClause<OMPThreadLimitClause>();
```
Move this out of outer `if`

alexey-bataev commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -9859,7 +9859,18 @@ void CGOpenMPRuntime::emitTargetCall(
+        emitTeamsCall(CGF, D, D.getBeginLoc(), OutlinedFn, CapturedVars);
+      }
+      else
+        emitOutlinedFunctionCall(CGF, D.getBeginLoc(), OutlinedFn, CapturedVars);
```
Enclose in braces

Commit:
Setting thread_limit icv

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -5119,6 +5119,13 @@ void CodeGenFunction::EmitOMPTargetTaskBasedDirective(
   IntegerLiteral IfCond(getContext(), TrueOrFalse,
                         getContext().getIntTypeForBitwidth(32, /*Signed=*/0),
                         SourceLocation());
+  if (S.getDirectiveKind() == OMPD_target && S.getSingleClause<OMPThreadLimitClause>()) {
+    const auto *TL = S.getSingleClause<OMPThreadLimitClause>();
+    if (TL) {
+      const Expr *ThreadLimit = TL ? TL->getThreadLimit() : nullptr;
+      CGM.getOpenMPRuntime().emitThreadLimitClause(*this, ThreadLimit, S.getBeginLoc());
+    }
+  }
```
```
  if (const auto *TL = S.getSingleClause<OMPThreadLimitClause>();;S.getDirectiveKind() == OMPD_target && TL) {
    const Expr *ThreadLimit = TL->getThreadLimit();
    CGM.getOpenMPRuntime().emitThreadLimitClause(*this, ThreadLimit, S.getBeginLoc());
  }
```

alexey-bataev commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -9837,7 +9837,8 @@ void CGOpenMPRuntime::emitTargetCall(
   const bool RequiresOuterTask = D.hasClausesOfKind<OMPDependClause>() ||
                                  D.hasClausesOfKind<OMPNowaitClause>() ||
-                                 D.hasClausesOfKind<OMPInReductionClause>();
+                                 D.hasClausesOfKind<OMPInReductionClause>() ||
+                                 D.hasClausesOfKind<OMPThreadLimitClause>();
```
Why do you need this?

sandeepkosuri replied:
This was to wrap the target region with a `task`. With this runtime implementation, I was hoping to set the `thread_limit` icv for this outer `task`.

alexey-bataev replied:
Hmm, should this thread limit server for inner tasks, not for the outer around target region? If yes, then this is just a waste of resources or does not work as intended.

sandeepkosuri replied:
By enclosing the entire target region with a task, we can manipulate the data env, hence the thread_limit icv as well for the entire target region. This method could set the thread_limit for all the constructs inside the target region in one shot instead of tracking and setting the limit for them individually.
If it's still a bad practice to do this, are there any other ways to set thread_limit for the entire target region, that I can explore ?

alexey-bataev replied:
Ye, I think it is not the best decision. Better to rework libomp/libomptarget to handle this parameter directly.

sandeepkosuri replied:
Can you please explain a bit more about the idea you are suggesting ?

@dreachem, do you have any comments ?

sandeepkosuri replied:
Currently I am trying to make the case of "executing the offloading function on the host itself" work, not a real target offloading case. So I am not touching the libomp/libomptarget

alexey-bataev replied:
Why do you need outer task for target? Does spec require it? Or just implementation?

sandeepkosuri replied:
Just implementation. This way I can set the thread_limit for the entire task instead of setting it to individual parallel directives which occur in target region.

Commit:
Partial thread_limit support for target directive (host execution only)

Commit:
Shifted codegen changes to EmitOMPTargetTaskBasedDirective() & code refactoring

Commit:
fixed check-openmp test failures

Commit:
placed version checks

Commit:
Shifted codegen changes to EmitOMPTargetTaskBasedDirective() and started using a different ICV to keep track of thread_limit

alexey-bataev commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -10653,6 +10656,24 @@ void CGOpenMPRuntime::emitNumTeamsClause(CodeGenFunction &CGF,
+                                      CGF.CGM.Int32Ty, /* isSigned = */ true)
+          : CGF.Builder.getInt32(0);
+
+  // Build call __kmpc_thread_limit(&loc, global_tid, thread_limit)
```
kmpc_set_thread_limit

sandeepkosuri replied:
Thanks for noticing. I fixed it.

Commit:
Edited omp_get_thread_limit() to support new thread_limit icv and fixed some lit test case issues.

Commit:
Merge branch 'thread_limit_2' of https://github.com/sandeepkosuri/llvm-project into thread_limit_2

sandeepkosuri commented on openmp/runtime/src/kmp_csupport.cpp:
```diff
@@ -381,6 +381,24 @@ void __kmpc_push_num_teams(ident_t *loc, kmp_int32 global_tid,
+current task
+
+Set the thread_limit for the current task
+This call is there to support `thread_limit` clause on the `target` construct
```
Yes, it does. I have just added this scenario in my LIT test case. Thank you for this input.

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -5110,6 +5110,16 @@ void CodeGenFunction::EmitOMPTargetTaskBasedDirective(
+        S.getSingleClause<OMPThreadLimitClause>()) {
+      const auto *TL = S.getSingleClause<OMPThreadLimitClause>();
+      if (TL) {
+        const Expr *ThreadLimit = TL ? TL->getThreadLimit() : nullptr;
```
You don't need extra checks here, you already checked that TL is not nullptr in the if above.

sandeepkosuri replied:
Sure, I will remove this extra check.

Commit:
added lit test

alexey-bataev commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -9837,7 +9837,10 @@ void CGOpenMPRuntime::emitTargetCall(
 
   const bool RequiresOuterTask = D.hasClausesOfKind<OMPDependClause>() ||
```
Do you still need this?

sandeepkosuri replied:
I'm sorry, are you asking about this :
```
const bool RequiresOuterTask = D.hasClausesOfKind<OMPDependClause>() ||
```
or this :

```
D.hasClausesOfKind<OMPInReductionClause>() ||
(D.getDirectiveKind() == OMPD_target &&
 D.hasClausesOfKind<OMPThreadLimitClause>() &&
 CGM.getLangOpts().OpenMP >= 51);
```

The first one is not part of my changes.
The second one is still required so that I can manipulate icv of task and apply thread_limit all over target region (technically task region) in one shot.

soumitrachatterjee commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -9837,7 +9837,10 @@ void CGOpenMPRuntime::emitTargetCall(
+                                 D.hasClausesOfKind<OMPInReductionClause>() ||
+                                 (D.getDirectiveKind() == OMPD_target &&
+                                  D.hasClausesOfKind<OMPThreadLimitClause>() &&
+                                  CGM.getLangOpts().OpenMP >= 51);
```
It might be a tad more efficient for pre-5.1 compilations if you put the check `CGM.getLangOpts().OpenMP >= 51` first in the sequence.

soumitrachatterjee commented on clang/lib/CodeGen/CGOpenMPRuntime.cpp:
```diff
@@ -10653,6 +10656,24 @@ void CGOpenMPRuntime::emitNumTeamsClause(CodeGenFunction &CGF,
+  llvm::Value *ThreadLimitVal =
+      ThreadLimit
+          ? CGF.Builder.CreateIntCast(CGF.EmitScalarExpr(ThreadLimit),
+                                      CGF.CGM.Int32Ty, /* isSigned = */ true)
```
Is it required to be a signed integer (negative values don't really make sense as a thread limit)?

soumitrachatterjee commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -5110,6 +5110,16 @@ void CodeGenFunction::EmitOMPTargetTaskBasedDirective(
 
     Action.Enter(CGF);
     OMPLexicalScope LexScope(CGF, S, OMPD_task, /*EmitPreInitStmt=*/false);
+    if (CGF.CGM.getLangOpts().OpenMP >= 51 &&
+        S.getDirectiveKind() == OMPD_target &&
+        S.getSingleClause<OMPThreadLimitClause>()) {
+      const auto *TL = S.getSingleClause<OMPThreadLimitClause>();
+      if (TL) {
+        const Expr *ThreadLimit = TL ? TL->getThreadLimit() : nullptr;
+        CGF.CGM.getOpenMPRuntime().emitThreadLimitClause(CGF, ThreadLimit,
+                                                         S.getBeginLoc());
+      }
+    }
```
```suggestion
    if (CGF.CGM.getLangOpts().OpenMP >= 51 &&
        S.getDirectiveKind() == OMPD_target &&
        S.getSingleClause<OMPThreadLimitClause>()) {
      CGF.CGM.getOpenMPRuntime().emitThreadLimitClause(CGF, 
                                                         S.getSingleClause<OMPThreadLimitClause>()->getThreadLimit(),
                                                         S.getBeginLoc());
    }
```

Commit:
Improved LIT test case and added comments
